File: Tuning/sample_code/02_timer-res/src/cpp/main.cpp
// Copyright 2024 Research Organization for Information Science and Technology 
//----------------------------------------------------------------------
//  Check resolution of your timer 
//  Author:      Yukihiro Ota (yota@rist.or.jp)
//  Description: The original idea comes from Wadleigh and Crawford,
//               "Software Optimization for HPC: Creating Faster 
//                Applications" (2000) pp.136-138 
//  Extended to compare the nominal resolution reported by the API
//  (std::chrono::steady_clock::period and CLOCKS_PER_SEC) with the
//  resolution measured empirically by the iteration loop.
//----------------------------------------------------------------------
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <ctime>

int func ( const int nn ) ;

//----------------------------------------------------------------------
//  main                                                              
//----------------------------------------------------------------------
int main ( int argc, char* argv[] )
{
  int nn ;
  int nn_max = 10000000 ; // safety limit to prevent infinite loop
  double t ;
  double res_api ;

  printf ("--------------------------------------------------------\n");

  // check resolution of wallclock timer with steady_clock
  printf ("[Wallclock timer]\n") ;
  res_api = static_cast<double>(std::chrono::steady_clock::period::num) /
            static_cast<double>(std::chrono::steady_clock::period::den) ;
  printf (" [API]      nominal resolution   = %17.9f sec.\n", res_api) ;

  nn = 0 ;
  t = -1.0 ;
  while ( t <= 0.0 && nn < nn_max ) {
    nn++ ;
    const auto elp1 = std::chrono::steady_clock::now();
    func (nn);
    const auto elp2 = std::chrono::steady_clock::now();
    const std::chrono::duration<double> elapsed = elp2 - elp1;
    t = static_cast<double>(elapsed.count());
  }

  if ( nn >= nn_max ) {
    printf ("Warning: wallclock timer resolution could not be determined.\n") ;
  } else {
    printf (" [Measured] observed resolution  = %17.9f sec. (nn = %d iterations)\n", t, nn) ;
    printf (" ratio (measured / API)          = %13.2f\n", t / res_api) ;
  }

  printf ("--------------------------------------------------------\n");

  // check resolution of cpu timer with std::clock
  printf ("[CPU timer]\n") ;
  res_api = 1.0 / static_cast<double>(CLOCKS_PER_SEC) ;
  printf (" [API]      nominal resolution   = %17.9f sec.\n", res_api) ;

  nn = 0 ;
  t = -1.0 ;
  while ( t <= 0.0 && nn < nn_max ) {
    nn++ ;
    const std::clock_t cpu1 = std::clock();
    func (nn);
    const std::clock_t cpu2 = std::clock();
    t = static_cast<double>(cpu2 - cpu1) / CLOCKS_PER_SEC ;
  }

  if ( nn >= nn_max ) {
    printf ("Warning: cpu timer resolution could not be determined.\n") ;
  } else {
    printf (" [Measured] observed resolution  = %17.9f sec. (nn = %d iterations)\n", t, nn) ;
    printf (" ratio (measured / API)          = %13.2f\n", t / res_api) ;
  }

  printf ("--------------------------------------------------------\n");

  /* finalization */
  return EXIT_SUCCESS ;
}
/*--------------------------------------------------------------------*/
/*  func                                                              */
/*--------------------------------------------------------------------*/
int func ( const int nn ) 
{
  int i, j ;
  i = 0 ;
  for ( j = 0 ; j < nn ; ++j ) i++ ;
  return i ;
}

File: Tuning/sample_code/02_timer-res/src/fortran/main.f90
! Copyright 2024 Research Organization for Information Science and Technology
!*************************************************************************
!  Check resolution of your timer
!  Author:      Yukihiro Ota (yota@rist.or.jp)
!  Description: The original idea comes from Wadleigh and Crawford,
!               "Software Optimization for HPC: Creating Faster 
!                Applications" (2000) pp.136-138 
!  Extended to compare the nominal resolution reported by
!  system_clock(count_rate=...) with the resolution measured
!  empirically by the iteration loop. cpu_time() has no API to query
!  its nominal resolution, so only the measured value is printed for
!  the CPU timer.
!*************************************************************************

!=========================================================================
!   module: mytype 
!=========================================================================
module mytype
  implicit none
  integer,parameter :: SP = kind(1.0)
  integer,parameter :: DP = selected_real_kind(2*precision(1.0_SP))
end module mytype
!=========================================================================
!   main 
!=========================================================================
program main

  use mytype,only: DP
  use iso_fortran_env, only: int64

  implicit none

  !! local var
  integer :: i, nn
  integer(int64) :: cnt1, cnt2, cnt_rate, cnt_max
  integer,parameter :: nn_max = 10000000

  real(kind=DP) :: tval, tval0, res_api
  real(kind=DP),parameter :: ZERO = 0.0_DP

  !! interface to function
  interface
    
    function func(nn)
      integer :: func
      integer,intent(in) :: nn
    end function func

  end interface

  write(6,'("----------------------------------------------------------")')

  !! check resolution of wallclock timer  
  write(6,'("[Wallclock timer]")')
  call system_clock (count_rate = cnt_rate)
  if ( cnt_rate > 0 ) then
    res_api = 1.0_DP / dble(cnt_rate)
  else
    res_api = -1.0_DP
  end if
  write(6,'(" [API]      nominal resolution   = ",1F17.9," sec.")') res_api

  tval = -1.0_DP 
  nn = 0
  do while ( tval .le. ZERO .and. nn .lt. nn_max )
    nn = nn + 1
    call system_clock (cnt1)
    i = func(nn)
    call system_clock (cnt2, cnt_rate, cnt_max)
    if ( cnt_rate > 0 ) then
      if ( cnt2 .ge. cnt1 ) then
         tval = (cnt2 - cnt1) / dble(cnt_rate)
      else
         tval = (cnt2 - cnt1 + cnt_max + 1) / dble(cnt_rate)
      end if
    end if
  end do

  if ( nn .ge. nn_max ) then
    write(6,'("Warning: wallclock timer resolution could not be determined.")')
  else
    write(6,'(" [Measured] observed resolution  = ",1F17.9," sec. (nn = ",1I0," iterations)")') tval, nn
    if ( res_api > ZERO ) then
      write(6,'(" ratio (measured / API)          = ",1F13.2)') tval / res_api
    end if
  end if

  write(6,'("----------------------------------------------------------")')

  !! check resolution of cpu timer  
  write(6,'("[CPU timer]")')
  write(6,'(" [API]      no nominal-resolution API is available for cpu_time()")')

  tval = -1.0_DP 
  nn = 0
  do while ( tval .le. ZERO .and. nn .lt. nn_max )
    nn = nn + 1
    call cpu_time (tval0)
    i = func(nn)
    call cpu_time (tval)
    tval = tval - tval0
  end do

  if ( nn .ge. nn_max ) then
    write(6,'("Warning: cpu timer resolution could not be determined.")')
  else
    write(6,'(" [Measured] observed resolution  = ",1F17.9," sec. (nn = ",1I0," iterations)")') tval, nn
  end if

  write(6,'("----------------------------------------------------------")')

  stop
end program main
!==========================================================================
!   func 
!==========================================================================
function func (nn)

  implicit none

  integer :: func
  integer,intent(in) :: nn

  integer :: j

  func = 0
  do j = 1, nn
    func = func + 1
  end do
end function func
